File: 269-alien-dictionary/alien-dictionary.cpp
class Solution {
public:
    unordered_map<char, vector<char>> mapper;
    unordered_map<char, bool> visited;
    vector<char> result;

    bool dfs(char i){
        if(visited.contains(i)) return visited[i];

        visited[i] = true;

        for(char j: mapper[i]){
            if (dfs(j)) return true;
        }

        visited[i] = false;
        result.push_back(i);
        return false;
    }

    string alienOrder(vector<string>& words) {
        
        for (string word: words){
            for (char l: word){
                mapper[l] = vector<char>();
            }
        }
        for (int i = 0; i < words.size() - 1; i++){
            string word1 = words[i];
            string word2 = words[i + 1];
            int mini = min(word1.length(), word2.length());
            if (word1.length() > word2.length() && word1.substr(0, mini) == word2.substr(0, mini)){
                return "";
            }
            for (int j = 0; j < mini; j++){
                if(word1[j] != word2[j]){
                    mapper[word1[j]].push_back(word2[j]);
                    break;
                }
            }
        }
        
        for (pair<char, vector<char>> temp: mapper){
            if (dfs(temp.first)){
                return "";
            }
        }

        reverse(result.begin(), result.end());
        return string(result.begin(), result.end());
    }
};

File: 269-alien-dictionary/alien-dictionary.py
class Solution:
    def alienOrder(self, words: List[str]) -> str:
        mapper = {l:[] for word in words for l in word}
        for i in range(len(words) - 1):
            word1 = words[i]
            word2 = words[i + 1]
            mid = min(len(word1), len(word2))
            if word1[:mid] == word2[:mid] and len(word1) > len(word2):
                return ""
            for j in range(mid):
                if word1[j] != word2[j]:
                    mapper[word1[j]].append(word2[j])
                    break
        
        result = []
        visited = {}
        def dfs(i):
            if i in visited:
                return visited[i]
            
            visited[i] = True
            for j in mapper[i]:
                if dfs(j):
                    return True
            visited[i] = False
            result.append(i)
        
        for i in mapper:
            if dfs(i):
                return ""
        
        result.reverse()

        return ''.join(result)